packages/blocks/src/lib/slam-map-2d/hooks/use-raycast-click.ts

```diff
@@ -0,0 +1,67 @@
+import * as THREE from 'three';
+import { useThrelte } from '@threlte/core';
+import { createEventDispatcher, onMount } from 'svelte';
+
+interface Events {
+  /**
+   * Fires when the user clicks on the canvas, and reports
+   * the intersection of a ray cast from the mouse coordinate
+   * to an infinite xy plane at z=0.
+   */
+  click: THREE.Vector3;
+}
+
+const EPSILON = 0.001;
+
+export const useRaycastClick = () => {
+  const { renderer, camera } = useThrelte();
+  const dispatch = createEventDispatcher<Events>();
+
+  const raycaster = new THREE.Raycaster();
+  const pointerDown = new THREE.Vector2();
+  const pointerUp = new THREE.Vector2();
+  const plane = new THREE.Plane();
+  plane.normal.set(0, 0, -1);
+
+  const normalizedDeviceCoordinates = (
+    x: number,
+    y: number,
+    vec2: THREE.Vector2
```

Comment:
> Even better, normalizeDeviceCoordinates should be an external function with unit tests to check its math

Good point, this is a common enough task, I'll extract & unit test.

> Could pointerDown and pointerUp be renamed something more obvious

Any suggestions? `pointerDownCoords`?

> Could it take only x and y and return the resulting vector

This is where conflicting principles I think will continue to clash in these PRs. When I'm in "real-time-rendering" mode when writing these components, my #1 scaling strategy is object pooling. Although performing GC on a simple struct-like on every click is negligible, these objects that need to be collected add up and the app's performance inevitably dies by a thousand cuts.

It's possible to mix and match strategies but I find that to be more confusing ultimately, since the threshold for when to pool and when not to pool is ambiguous, and profiling all cases can get quite tedious. What you're observing here is therefore a common pattern in Three.js or Babylon.js: objects are pooled for the longest possible / reasonable lifetime. Below I'm dispatching a new `Vector3` for every click, but that's only because it's dispatched outside of this component and I have no control over what the consumer will do with it.

Reply:
To be clear, I'm not requesting this moves away from object mutation, I'm requesting that the computation is a separate function from the mutation, so that the logic can be tested

Reply:
Got it, in that case, test added!